File: src/tp03/ej09/Triangle.java
package tp03.ej09;

public class Triangle extends Figure {
    private Point corner1, corner2, corner3;

    Triangle(Point corner1, Point corner2, Point corner3) {
        this.corner1 = corner1;
        this.corner2 = corner2;
        this.corner3 = corner3;
    }

    public double perimeter() {
        return corner1.distanceTo(corner2) + corner2.distanceTo(corner3) + corner3.distanceTo(corner1);
    }

    public double area() {
        double p = perimeter()/2;
        return Math.sqrt(p * (p - corner1.distanceTo(corner2)) * (p - corner2.distanceTo(corner3)) * (p - corner3.distanceTo(corner1)));
    }

    @Override
    public String toString() {
        return "Triangulo [ " + corner1 + " , " + corner2 + " , " + corner3 + " ]";
    }
}

File: src/tp08/ej09/triangle.rb
class Triangle < Figure
  attr_reader :point1, :point2, :point3

  def initialize(point1, point2, point3)
    @point1 = point1
    @point2 = point2
    @point3 = point3
  end

  def perimeter
    side1 = @point1.distance(@point2)
    side2 = @point2.distance(@point3)
    side3 = @point3.distance(@point1)
    side1 + side2 + side3
  end

  def area
    # Using Heron's formula to calculate the area
    a = @point1.distance(@point2)
    b = @point2.distance(@point3)
    c = @point3.distance(@point1)
    s = (a + b + c) / 2.0
    Math.sqrt(s * (s - a) * (s - b) * (s - c))
  end

  def ==(other)
    return false unless other.is_a? Triangle
    @point1 == other.point1 && @point2 == other.point2 && @point3 == other.point3
  end

  def to_s
    "Traingle [ #{@point1} #{@point2} #{@point3} ]"
  end
end
